File: c++/67_Add_Binary.cpp
#include <string>
#include <iostream>
using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
        int i = a.size() - 1;
        int j = b.size() - 1;
        int carry = 0;
        string result = "";

        result.reserve(max(a.size(), b.size()) + 1);

        while (i >= 0 || j >= 0 || carry) {
            // we need to convert the char to int by subtracting '0'
            int digit_a = (i >= 0) ? a[i] - '0' : 0;
            int digit_b = (j >= 0) ? b[j] - '0' : 0;

            int sum = digit_a + digit_b + carry;

            result.push_back((sum % 2) + '0');
            carry = sum / 2;

            i--;
            j--;
        }

        reverse(result.begin(), result.end());
        return result;
    }
};

int main() {
    Solution sol;

    cout << sol.addBinary("11", "1") << endl;        // expected: 100
    cout << sol.addBinary("1010", "1011") << endl;   // expected: 10101
    cout << sol.addBinary("0", "0") << endl;         // expected: 0
    cout << sol.addBinary("1", "1") << endl;         // expected: 10
    cout << sol.addBinary("111", "111") << endl;     // expected: 1110

    return 0;
}

File: Python/67_Add_Binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i = len(a) - 1
        j = len(b) - 1
        carry = 0
        result = ""

        # result.reserve(max(len(a), len(b)) + 1)
        #  this only applies to C++ to optimize memory allocation, in Python strings are immutable so we will just concatenate and reverse at the end
        while i >= 0 or j >= 0 or carry:
            digit_a = (int(a[i]) if i >= 0 else 0)
            digit_b = (int(b[j]) if j >= 0 else 0)

            sum = digit_a + digit_b + carry

            result += str(sum % 2)
            carry = sum // 2

            i -= 1
            j -= 1
        
        return result[::-1]